Add the ability for the IPCTestingAPI to report when a sync stream message fails to be deserialized correctly

gavin-apple opened the pull request:
#### 56b0d92bb3fe7fbbdd946601c98e2e5b5b68fb75
<pre>
Add the ability for the IPCTestingAPI to report when a sync stream message fails to be deserialized correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=264810">https://bugs.webkit.org/show_bug.cgi?id=264810</a>
<a href="https://rdar.apple.com/118391408">rdar://118391408</a>

Reviewed by Alex Christensen.

Add the ability for the IPCTestingAPI to report when a sync stream connection
message fails to be received correctly. The existing behaviour is that no reply
is returned for failed messages. This change records the decoder validity
on failure into a MessageSyncReply so that the IPCTestingAPI can raise
an exception.

* LayoutTests/ipc/ignoreinvalidmessagesfortesting-stream-connection-expected.txt: Added.
* LayoutTests/ipc/ignoreinvalidmessagesfortesting-stream-connection.html: Added.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::createRenderingBackend):
(WebKit::GPUConnectionToWebProcess::createGraphicsContextGL):
(WebKit::GPUConnectionToWebProcess::createRemoteGPU):
* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::handleMessageSynchronous):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::tryCreate):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage):
(IPC::StreamServerConnection::sendDeserializationErrorSyncReply):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::IPCStreamTester):
(WebKit::IPCStreamTester::syncMessageEmptyReply):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/270892@main">https://commits.webkit.org/270892@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ee24c275c80ef000add49e35b6fdca0f3dcfbed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26399 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28545 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24162 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26736 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6882 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24152 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26659 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3910 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22701 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3424 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29096 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24124 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24077 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29732 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3472 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1696 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27626 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4898 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6422 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3797 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gavin-apple commented on Source/WebKit/Shared/IPCStreamTester.cpp:
```diff
@@ -93,6 +93,12 @@ void IPCStreamTester::syncMessageReturningSharedMemory1(uint32_t byteCount, Comp
 
+void IPCStreamTester::syncMessageEmptyReply(uint32_t value, CompletionHandler<void()>&& completionHandler)
+{
+    UNUSED_PARAM(value);
```
I need value so I can trigger a deserialisation error when it's excluded

achristensen07 replied:
You need this function to have a uint32_t parameter, but you don't need to name it "value"
...(uint32_t, CompletionHandler<...

achristensen07 commented on Source/WebKit/Platform/IPC/HandleMessage.h:
```diff
@@ -288,8 +288,10 @@ void handleMessageSynchronous(StreamServerConnection& connection, Decoder& decod
     auto arguments = decoder.decode<typename MessageType::Arguments>();
-    if (UNLIKELY(!arguments))
+    if (UNLIKELY(!arguments)) {
+        connection.sendDeserializationErrorSyncReply(syncRequestID);
```
It seems like this should be protected by #if ENABLE(IPC_TESTING_API)

gavin-apple committed:
Add the ability for the IPCTestingAPI to report when a sync stream message fails to be deserialized correctly
https://bugs.webkit.org/show_bug.cgi?id=264810
rdar://118391408

Reviewed by Alex Christensen.

Add the ability for the IPCTestingAPI to report when a sync stream connection
message fails to be received correctly. The existing behaviour is that no reply
is returned for failed messages. This change records the decoder validity
on failure into a MessageSyncReply so that the IPCTestingAPI can raise
an exception.

* LayoutTests/ipc/ignoreinvalidmessagesfortesting-stream-connection-expected.txt: Added.
* LayoutTests/ipc/ignoreinvalidmessagesfortesting-stream-connection.html: Added.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::createRenderingBackend):
(WebKit::GPUConnectionToWebProcess::createGraphicsContextGL):
(WebKit::GPUConnectionToWebProcess::createRemoteGPU):
* Source/WebKit/Platform/IPC/HandleMessage.h:
(IPC::handleMessageSynchronous):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::tryCreate):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage):
(IPC::StreamServerConnection::sendDeserializationErrorSyncReply):
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
* Source/WebKit/Shared/IPCStreamTester.cpp:
(WebKit::IPCStreamTester::IPCStreamTester):
(WebKit::IPCStreamTester::syncMessageEmptyReply):
* Source/WebKit/Shared/IPCStreamTester.h:
* Source/WebKit/Shared/IPCStreamTester.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/270892@main
